onnxruntime/contrib_ops/cuda/quantization/dequantize_blockwise_bnb4.cu

```diff
@@ -119,11 +126,36 @@ template Status DequantizeBnb4<half>(
     const half* quant_map,
     half* output,
     const uint8_t* quant_data,
-    const half *absmax,
+    const half* absmax,
     int block_size,
     int numel,
     cudaStream_t stream);
 
+template <>
+Status DequantizeBnb4<BFloat16>(
+    const BFloat16* quant_map,
+    BFloat16* output,
+    const uint8_t* quant_data,
+    const BFloat16* absmax,
+    int block_size,
+    int numel,
+    cudaStream_t stream) {
+  #if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 800
```

Comment:
what is the reason we need handle >=800 differently? I did not see other kernels do such to support bfloat16.

Reply:
The onnxruntime built-in `BFloat16` type does compute in `float` https://github.com/microsoft/onnxruntime/blob/a5537f2f563d4975c7e6121a7eb260bbbfd9455a/onnxruntime/core/providers/rocm/cu_inc/common.cuh#L17 
While profiling, I found that this is slower than using the native bfloat16 type in cuda (only works >= 800). Also, we won't have accumulate the deviations coming from casting to bfloat16 after each multiplication. 
This also matches the behavior in the bitsandbytes kernel https://github.com/TimDettmers/bitsandbytes/blob/main/csrc/kernels.cu#L3590